PowerEditor/visual.net/notepadPlus.vcxproj

```diff
@@ -122,6 +122,7 @@
       <SubSystem>Windows</SubSystem>
       <TargetMachine>MachineX86</TargetMachine>
       <MinimumRequiredVersion>5.01</MinimumRequiredVersion>
+      <AdditionalLibraryDirectories>..\..\scintilla\bin;C:\Libraries\boost_1_69_0\lib32-msvc-14.1\</AdditionalLibraryDirectories>

```

Comment:
I think the preferable solution would be:
https://social.msdn.microsoft.com/Forums/vstudio/en-US/c5fa62ac-e84f-4d61-a9bf-eb8136305393/creating-a-static-libraries-that-depends-on-other-libraries?forum=vclanguage
- modify the build of libscilexer.lib to already contain the dependent libs of boost, imm32, msimg32.lib within the static one

Also possible:
See https://social.msdn.microsoft.com/Forums/en-US/2b2e070c-2918-49ed-a64e-7ef413272e98/how-to-include-vc-directory-paths-defined-in-microsoftcppuserprop-file-in-msbuild-command-line?forum=msbuild
- We could add a prop file and define there the boost path
- add some default path which is working for manual usage and configure appveyor via commandline

- use nuget packages for boost within N++

Reply:
@chcg 
Ideally it would be great to add Scintilla project  into the solution and the compiled boost regexpr static lib into repo, so any dev can pull the repo and build Notepad++ with whole features via VS2017(VS2019 in the future)'s `Build Solution` command.

But so far I'm trying to find a middle term solution for both dev (like you & me) and Appveyor build system. So `C:\Libraries\boost_1_69_0\lib32-msvc-14.1\` cannot be used, because it's only for build system. The middle term solution for me is this PR with a empty boost lib directory (without version number) in repository (a placeholder dummy file inside it), plus the additional build instruction to tell users to put the compiled boost pcre lib into this directory.
In this way we have the solution now for both dev & build environment, and users can tests new binary immediately.
What do you think?

Reply:
@donho Did I get you right that a only VS solution (N++ with scilexer) would be the goal?
In this case we had already something like that, see below. The boost stuff in that case is coming from a nuget package (https://github.com/sergey-shandar/getboost) which contains prebuilt binaries, taken from https://dl.bintray.com/boostorg/release/1.72.0/binaries/ and is automatically downloaded by VS studio. This could be also easily adapted for the static build.

What do you think about "Unicode Release" vs. just "Release"

> Revision: eba913d887726b25205356bb971f6eadc207033a
> Author: Christian Grasser <christian.grasser@live.de>
> Date: 13.03.2017 21:28:02
> Message:
> Scintilla Namespace
> 
> - corrected missing scintilla namespaces
> - activated usage of scintilla namespace in nmakefile and vcxproj
> 
> Closes #3033
> ----
> Modified: scintilla/boostregex/AnsiDocumentIterator.h
> Modified: scintilla/boostregex/UTF8DocumentIterator.cxx
> Modified: scintilla/boostregex/UTF8DocumentIterator.h
> Modified: scintilla/lexers/LexObjC.cxx
> Modified: scintilla/lexers/LexSearchResult.cxx
> Modified: scintilla/win32/SciLexer.vcxproj
> Modified: scintilla/win32/scintilla.mak
> 
> Revision: fdd2dbc21fb0de8f1ed90cda11f5169d20258c81
> Author: Christian Grasser <christian.grasser@live.de>
> Date: 11.06.2015 17:37:54
> Message:
> - add npp boostregex dir/sources and define SCI_OWNREGEX
> - add boost via nuget package
> ----
> Modified: scintilla/win32/SciLexer.vcxproj
> Added: scintilla/win32/packages.config
> 
> Revision: a6e0dd9a533187b7437f0596918489cfd427df74
> Author: Christian Grasser <christian.grasser@live.de>
> Date: 11.06.2015 11:31:59
> Message:
> adapted scintilla vs project to notepad++ naming of build configurations from
> Debug -> Unicode Debug and Release -> Unicode Release, to hav a consistent look in VS solution
> Used same
> ToolsVersion="12.0"
> and
> <PlatformToolset>v120_xp</PlatformToolset>
> ----
> Modified: scintilla/win32/SciLexer.vcxproj

Reply:
> Ideally it would be great to add Scintilla project into the solution and the compiled boost regexpr static lib into repo, so any dev can pull the repo and build Notepad++ with whole features via VS2017(VS2019 in the future)'s `Build Solution` command.

Just an idea. In theory, CMake could be used for that without the need of a project/solution file.
I think, it could allow building the entire N++ in "one click". 
It should handle any subprojects whether they're using Make or anything else.
It is integrated right into Visual Studio: https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-160#ide-integration
Adding a NuGet package to the CMake setup or any other package (e.g. vcpkg) should be relatively easy.
I can elaborate on that with a full list of pros/cons in a separate issue, if anybody is interested.

Reply:
@mere-human There is already a starting point for https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/CMakeLists.txt, but not used as a first class citizen and not part of the automatic check via appveyor, so likely to be outdated currently. Scintilla is currently not available as cmake build target, but that should be no big issue to create one. I think it just makes sense if it is kept up to date with the other build options.

Reply:
@chcg 
> Scintilla is currently not available as cmake build target, but that should be no big issue to create one. 

I just don't see why we should create CMakefile.txt for Scintilla project.

@mere-human 
> Just an idea. In theory, CMake could be used for that without the need of a project/solution file.

It's not because a solution is popular that we should use it.
CMake is cool for a cross-platform project. For a Windows-only application like Notepad++, it's pointless to me to use such solution.

Reply:
@chcg 
> Did I get you right that a only VS solution (N++ with scilexer) would be the goal?

Yes.

> In this case we had already something like that, see below. The boost stuff in that case is coming from a nuget package (https://github.com/sergey-shandar/getboost) which contains prebuilt binaries, taken from https://dl.bintray.com/boostorg/release/1.72.0/binaries/ and is automatically downloaded by VS studio. This could be also easily adapted for the static build.

If nuget works for both devs and Appveyor environment, the YES!

Reply:
> What do you think about "Unicode Release" vs. just "Release"

While it's possible to compile a project in ANSI, we keep "Unicode" to distinguish Unicode build from ANSI build.

Reply:
My thinking about having "Unicode" in the name was in relation to the recently published build-for-test wiki instructions (which may need tweaking after the build flux settles down).  It may confuse people into thinking that what is built can only be used on unicode (e.g. UTF8) files and not for "ANSI" files.  But it is a small matter, and I certainly have no objection to keeping Unicode in the name.

Reply:
Indeed, in the generated binary name "Unicode" is not necessary and it can be removed.

Reply:
Adapted the build VS solution to use nuget for boost regex. 
Until now I didn't manage to restrict the boost dependency to just the libscilexer.lib and there seems to be no good option to reference the needed static regex lib. This requires the nuget package also in the n++ vcxproj files.
The binary size seems to be bigger than with the build by the make script. 
Changed the appveyor.yml to just have Release and Debug in the artefacts name.

Reply:
>  This requires the nuget package also in the n++ vcxproj files.

In this case, I prefer to include pre-built boost lib.

Reply:
The nuget regex package contains also pre-built boost libs. And the same issue appears on using the ones from appveyor, so we remain with the problem to add an additional path to the boost libs into the N++ vcxproj file. I will try to find out why it is not sufficient to have the static lib be part of libscilexer.lib. This is working for e.g. imm32.lib.

Reply:
@chcg 
Sorry I wasn't clear. By "include pre-built boost lib", I meant do without nuget package since you said "This requires the nuget package also in the n++ vcxproj files.".

My vision to the future solution is:
```
Solution notepadPlus (notepadPlus.sln)
    Notepad++ (notepadPlus.vcxproj) - generate notepad++.exe (include libscilexer.lib)
    Scintilla (scintilla/win32/libSciLexer.vcxproj) - generate libscilexer.lib (include boostRegexpr.lib)
```
`boostRegexpr.lib` could be pre-built and included in the repository. In this way we have complete chain to build `notepad++.exe`, Building Notepad++ will be much easier for whomever - it needs only to fork the project locally, launch the solution then "Build the solution" or "Build Notepad++" command.

What do you think?

Reply:
@donho There are interesting things ongoing with boost 1.76 beta1, see https://www.boost.org/users/history/version_1_76_0.html. Regex will be changed to be a header only version.

> 
> Regex:
> 
>     Regex is now header only except in C++03 mode.
>     Support for C++03 is now deprecated.
>     The library can now be used "standalone" without the rest of Boost being present.

So maybe that will make it easier to use it for use. In the future.

Additionally there is already https://www.boost.org/doc/libs/1_75_0/doc/html/xpressive.html available, which is also header only and might be worth a look, if updated regex doesn't evolve in the right direction.

For now I will try to modify the PR and add the binary libs directly into git. This would be
libboost_regex-vc141-mt-s-x32-1_72.lib 12 MB
libboost_regex-vc141-mt-s-x64-1_72.lib 16 MB

I hope I got your proposal right.

Reply:
> The library can now be used "standalone" without the rest of Boost being present.

If that does mean boost regexpr contains only it-self, why don't we include it in the repository directly? 

```
Solution notepadPlus (notepadPlus.sln)
    Notepad++ (notepadPlus.vcxproj) - generate notepad++.exe (include libscilexer.lib)
    Scintilla (scintilla/win32/libSciLexer.vcxproj) - generate libscilexer.lib (include boostRegexpr.lib)
    BoostRegExpr (BoostRegExpr/boostRegExpr.vcxproj) generate boostRegexpr.lib
```

I hate git sub-Module and don't like the idea of Nuget - my goal is make build process as simple as possible so a grand-mom or a 8 year-old kid can build Notepad++ without RTFM. And with such simplicity, the binary generated by Appveyor will be near the binary of release - that's a great gain IMO.

Reply:
@donho Just spotted the boost beta today, so I didn't have a closer look on it yet. So it is unclear to me what "standalone" means exactly.

Maybe you like subtree more than submodule:
https://blog.developer.atlassian.com/the-power-of-git-subtree/?_ga=2-71978451-1385799339-1568044055-1068396449-1567112770

As long as you are staying within VisualStudio nuget packages are hidden behind the scenes for the "end user" and just pressing F7 is doing the build + checkin that the package exists. So not too far away from your expectation.